lib/os/onoff.c

```diff
@@ -0,0 +1,503 @@
+	} else if (notify) {
+		srv->refs += 1;
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (start) {
+		__ASSERT_NO_MSG(srv->start != NULL);
+		srv->start(srv, onoff_start_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res)
+{
+	bool notify_clients = false;
+	int client_res = res;
+	bool start = false;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+	struct onoff_client *releaser = srv->releaser;
+
+	/* If the stop operation failed log an error and leave the
+	 * rest of the state in place.
+	 *
+	 * If it succeeded remove the last reference and transition to
+	 * off.
+	 *
+	 * In either case remove the last reference, and notify all
+	 * waiting clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+		notify_clients = true;
+		sys_slist_init(&srv->clients);
+	} else if (sys_slist_is_empty(&clients)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+	} else if (k_is_in_isr()
+		   && ((srv->flags
+			& ONOFF_SERVICE_ISR_CALLABLE_START)
+		       == 0U)) {
+		notify_clients = true;
+		client_res = -EWOULDBLOCK;
+	} else {
+		start = true;
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+	}
+
+	srv->releaser = NULL;
+	srv->refs -= 1U;
+	__ASSERT_NO_MSG(srv->refs == 0);
+
+	k_spin_unlock(&srv->lock, key);
+
+	/* Notify the releaser.  If there was an error, notify any
+	 * pending requests; otherwise if there are pending requests
+	 * start the transition to ON.
+	 */
+	notify_one(srv, releaser, res);
+	if (notify_clients) {
+		notify_all(srv, &clients, client_res);
+	} else if (start) {
+		srv->start(srv, onoff_start_notify);
+	}
+}
+
+int z_impl_onoff_release(struct onoff_service *srv,
+			 struct onoff_client *cli)
+{
+	bool stop = false;      /* invoke stop transition */
+	bool notify = false;    /* do client notification */
+	bool no_wait = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv != 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_ON:
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
+			break;
+		}
+
+		/* Reject if in isr but stop not isr-callable */
+		if (k_is_in_isr()
+		    && ((srv->flags & ONOFF_SERVICE_ISR_CALLABLE_STOP)
+			== 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Reject if in no-wait condition and stop could
+		 * suspend.
+		 */
+		if (no_wait
+		    && ((srv->flags & ONOFF_SERVICE_SUSPENDABLE_STOP)
+			!= 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		stop = true;
+
+		set_service_state(srv, SERVICE_STATE_TO_OFF);
+		srv->releaser = cli;
+		rv = 2;
+
+		break;
+	case SERVICE_STATE_TO_ON:
+		rv = -EBUSY;
```

Comment:
this can be quite error prone. Let say that module A started the clock but before it got started clock is no longer needed. Module A releases the clock but gets `-EBUSY`. Now, there must be a logic in the notification callback which on successful clock starting needs to check the status which must say something like: disable but failed to release the clock. Based on that status release must be called again.

An alternative would be to decrement the counter, remove client from list and return success. Then in `onoff_start_notify` trigger stop if `ref_cnt == 0`. Return `-EBUSY` only if client is not found in the list. That would simplify early service releasing.

Reply:
Requesting a start, then requesting a stop before the start completes, is functionally a "cancel in-progress operation" request.  The API currently doesn't support that, but there are reasons why it should, and I'll take a look at that later.  It would be a separate function.

Reply:
@pabigot do you plan to add this feature in this PR or as next step once this get merged?

Reply:
It's already implemented.  `onoff_cancel()`.